runtime/src/identity.rs

```diff
@@ -380,6 +396,136 @@ decl_module! {
+        ///
+        /// # Failure
+        /// Only called by master key owner.
+        ///
+        /// # Arguments
+        /// * `did` caller's DID
+        /// * `new_key` master key replacing current key
+        pub fn change_master_key(origin, did: IdentityId, new_key: Key) -> Result {
+            let sender = ensure_signed(origin)?;
+            let master_key = Key::try_from(sender.encode())?;
+            let _grants_checked = Self::grant_check_only_master_key(&master_key, did)?;
+
+            ensure!(!<MasterKeyRotation>::exists(did), "a key rotation process already in progress");
+
+            let rotation = MasterKeyRotationPrefs {
+                new_key,
+                target_accepted: false,
+                kyc_verified: false
+            };
+
+            <MasterKeyRotation>::insert(did, rotation);
+
+            Self::add_auth(Signer::Key(master_key), Signer::Key(new_key), AuthorizationData::RotateMasterKey(master_key, did), None);
+
+            Self::deposit_event(RawEvent::MasterKeyChangeInitiated(did, new_key));
+
+            Ok(())
+        }
+
+        /// Call this with the new master key. By invoking this method, caller accepts authorization
+        /// `auth_id` with the new master key. If a KYC service provider approved this change, master
+        /// key of the DidRecord is updated
+        ///
+        /// # Arguments
+        /// * `did` caller's DID
+        /// * `auth_id` AuthorizationID that corresponds to changing master key
+        pub fn accept_master_key(
+            origin,
+            did: IdentityId,
+            auth_id: u64
+        ) -> Result {
+            let new_key = Key::try_from(ensure_signed(origin)?.encode())?;
+            let new_signer = Signer::from(new_key);
+
+            ensure!(<Authorizations<T>>::exists((new_signer, auth_id)), "Invalid auth");
+            let auth = Self::authorizations((new_signer, auth_id));
+
+            let (current_master_key, did) = match auth.authorization_data {
+                AuthorizationData::RotateMasterKey(current_master_key, did) => (current_master_key, did),
+                _ => return Err("Invalid authorization")
+            };
+
+            ensure!(
+                <MasterKeyRotation>::exists(did),
+                "master key rotation request does not exist"
+            );
+
+            // If this call originated from the new master key, `sender_key` would be the new master key
+            if let Some(rotation) = <MasterKeyRotation>::get(did) {
+                <MasterKeyRotation>::mutate(did, |entry| {
+                    if let Some(rotation) = entry {
+                        rotation.target_accepted = true
+                    }
+                });
+
+                if rotation.kyc_verified {
+                    Self::rotate_master_key(did, rotation.new_key);
+                }
+
+                Self::consume_auth(Signer::from(current_master_key), new_signer, auth_id)?;
```

Comment:
auth should be consumed before changes are made (`rotate_master_key` is called). This is because the `consume_auth` function can revert if the auth has expired or is invalid or something like that. We don't want changes to happen when an invalid auth is provided.